### PR TITLE
libc: common: add support for iso c11 threads

### PR DESCRIPTION
**Pull request by cfriedt**
**Note: Compliance issues are false positives**

Capitalize on newly added support for dynamic thread stacks and the existing pthread support to provide an implementation of ISO C11 `<threads.h>`.

For more information about the ISO C standard, please see
https://www.iso.org/standard/74528.html

### CODE REVIEW

**Comment by cfriedt** [comment]
Screenshot 2023-07-24 at 9 51 17 PM

File: lib/libc/minimal/source/thrd/thrd.c
Comment (cfriedt) [inline lib/libc/minimal/source/thrd/thrd.c:51]: For consistent layering, this should be replaced with `pthread_yield()`, but it's effectively the same thing.

File: tests/lib/c_lib/prj.conf
Comment (cfriedt) [inline tests/lib/c_lib/prj.conf:8]: We'll likely need some additional Kconfig, dependencies, etc.

**Comment by cfriedt** [comment]
Just looking for some early feedback

File: lib/libc/minimal/source/thrd/thrd.c
Comment (ycsin) [reply lib/libc/minimal/source/thrd/thrd.c:51]: how about [`sched_yield`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/posix/sched.h#L40)?

Comment (keith-packard) [reply lib/libc/minimal/source/thrd/thrd.c:51]: I think cfriedt wanted to ensure that the whole c11 thread API was implemented in terms of standard pthreads APIs; poking through to `k_yield` exposes a gap in that layer.

**Review by keith-packard — COMMENTED** [review]
This is looking fine. Is there any reason to depend upon the C library thread.h? Or could Zephyr provide its own version of that file directly?

File: lib/libc/minimal/source/thrd/thrd.c
Comment (cfriedt) [reply lib/libc/minimal/source/thrd/thrd.c:51]: I did add https://github.com/zephyrproject-rtos/zephyr/issues/60757 but then closed it shortly after ycsin told me it was deprecated, and it's no longer on the opengroup site.

For reference
https://man7.org/linux/man-pages/man3/pthread_yield.3.html

I guess I don't have anything against adding `pthread_yield()` even if it is deprecated, but technically `sched_yield()` would work too.

**Comment by cfriedt** [comment]
> This is looking fine. Is there any reason to depend upon the C library thread.h? Or could Zephyr provide its own version of that file directly?

No reason to explicitly not make it common in Zephyr that I can think of - @keith-packard, were you thinking it might be better to put this into `lib/libc/common/`?

**Comment by keith-packard** [comment]
> No reason to explicitly not make it common in Zephyr that I can think of - @keith-packard, were you thinking it might be better to put this into `lib/libc/common/`?

It's almost something more common than libc/common -- shouldn't we always use this code, even if the C library provides alternatives?

**Comment by cfriedt** [comment]
Need #60964 and #60961 to be merged before this.

**Comment by cfriedt** [comment]
Fixed another regression in `pthread_mutex_timedlock()` (moved to a separate PR).

File: tests/lib/c_lib/src/test_thrd.c
Comment (cfriedt) [reply tests/lib/c_lib/src/test_thrd.c:152]: Fixed!

**Comment by cfriedt** [comment]
* added support for C11 mutexes

File: lib/libc/minimal/source/thrd/thrd.c
Comment (cfriedt) [reply lib/libc/minimal/source/thrd/thrd.c]: Fixed!

**Comment by cfriedt** [comment]
* filled-in `cnd` testsuite

**Comment by cfriedt** [comment]
* added thread-specific storage (tss) and `call_once()` support

**Comment by cfriedt** [comment]
Maybe I'm missing something, but it seems that there are a number of "Testsuite mismatch" errors.

Previously, it was possible to have multiple `ZTEST_SUITE()` declarations per actual testsuite. This is kind of nice, because even though some units are related as part of the same library, the tests may require different `setup()`, `teardown()`, `before()`, or `after()` functions, as well as different test `fixture`s.

```shell
INFO    - 1) tests/lib/c_lib/libraries.libc on mps2_an385 failed (Testsuite mismatch)
INFO    - 2) tests/lib/c_lib/libraries.libc.picolibc on mps2_an385 failed (Testsuite mismatch)
INFO    - 3) tests/lib/c_lib/libraries.libc.newlib on mps2_an385 failed (Testsuite mismatch)
INFO    - 4) tests/lib/c_lib/libraries.libc.newlib_nano on mps2_an385 failed (Testsuite mismatch)
INFO    - 5) tests/lib/c_lib/libraries.libc.minimal.strerror_table on mps2_an385 failed (Testsuite mismatch)
INFO    - 6) tests/lib/c_lib/libraries.libc.minimal.no_strerror_table on mps2_an385 failed (Testsuite mismatch)
```

In any cases, each of the testsuites above runs successfully with
```shell
west build -p auto -b mps2_an385 -t run -T <suite>
```

**Comment by cfriedt** [comment]
@nashif - it seems to me that there are some false negatives reported by twister. Can you please take a look?

https://github.com/zephyrproject-rtos/zephyr/actions/runs/5976216865/job/16213675832?pr=60759

File: include/threads.h
Comment (nashif) [inline include/threads.h:1]: inldude/threads.h? hmmm....

**Comment by nashif** [comment]
> @nashif - it seems to me that there are some false negatives reported by twister. All of these testsuites are passing FWICT, but there is a test suite mismatch (which I think is incorrect) that is getting reported
> 
> Can you please take a look?

I think I know what the issue is.. looking for a solution.

**Comment by nashif** [comment]
> Can you please take a look?

so, this is because we are not parsing the source files correctly, twister does not do recursive globbing and things in sub-dirs are not being scanned. Please move things like `src/thrd/cnd.c` to something line `src/thrd_cnd.c` for now. Will look into adding support for recursive scanning later on.

**Comment by cfriedt** [comment]
> @cfriedt so, this is because we are not parsing the source files correctly

@nashif - fixed in #61933

**Comment by cfriedt** [comment]
* rebased

**Comment by cfriedt** [comment]
* separated libc tests into common, thrd

File: include/threads.h
Comment (cfriedt) [reply include/threads.h:1]: i tried moving it to `libc/common/include` but then picolibc gave me an error.

File: lib/libc/common/source/thrd/mtx.c
Comment (dkalowsk) [inline lib/libc/common/source/thrd/mtx.c:36]: for 2 results, is it really worth having a switch or just a simple if/else?  I won't repeat the question on the other places where this happens as well.

Not really a requesting a change, just a minor question on code intent.  A `switch` implies that more conditional results will be of value in the future, but any of the other possible results for this don't matter, they are just failures.

**Review by dkalowsk — DISMISSED** [review]
A relatively minor question on style preference that shouldn't slow down merging.  After having done multiple re-reviews, I don't see anything on this that appears obviously wrong.  Nice job!

File: lib/libc/common/source/thrd/mtx.c
Comment (cfriedt) [reply lib/libc/common/source/thrd/mtx.c:36]: > A `switch` implies that more conditional results will be of value in the future, but any of the other possible results for this don't matter, they are just failures.

Yes, that's exactly it. I wanted to structure it so that errors could easily be more specialized or be corrected with ease.

Since you were curious though, here is a comparison 
https://godbolt.org/z/GzPv4nf5c

with default optimizations
```diff
diff -Nua /tmp/*.S
--- /tmp/with-if-stmt.S	2023-09-07 19:52:25
+++ /tmp/with-switch-stmt.S	2023-09-07 19:52:44
@@ -10,11 +10,11 @@
         ldr     x0, [sp, 40]
         bl      pthread_mutex_init
         cmp     w0, 0
-        beq     .L2
-        mov     w0, 2
+        bne     .L2
+        mov     w0, 0
         b       .L3
 .L2:
-        mov     w0, 0
+        mov     w0, 2
 .L3:
         ldp     x29, x30, [sp], 48
         ret

```

with `-O3`
```diff
<intentionally empty>
```

File: include/threads.h
Comment (nashif) [reply include/threads.h:1]: IMO this needs some discussion and broader review, putting this at this level is just a source for problems and confusion.

**Review by nashif — DISMISSED** [review]
location of new headers needs some more discussion.

**Comment by cfriedt** [comment]
Arch Meeting: 2023-09-12: not enough time to cover this topic.

**Comment by cfriedt** [comment]
I moved the `threads.h` header file to `lib/libc/common/include`.

@nashif - does that address your concerns?

The PicoLibC error I encountered is fixed.

Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.

However, I also wanted to go a step further and ensure that this would still be usable with external toolchains, such as GNUARM Embedded. Following the instructions [here](https://lindevs.com/install-arm-gnu-toolchain-on-ubuntu), as well as the usual [GSG](https://docs.zephyrproject.org/latest/develop/beyond-GSG.html#install-a-toolchain), the testsuite passes.
```shell
export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
export CROSS_COMPILE=/opt/gcc-arm-none-eabi/bin/arm-none-eabi-
west build -p auto -b mps2_an385 -t run tests/lib/c_lib/thrd
```

Also succeeds with LLVM
```shell
export ZEPHYR_TOOLCHAIN_VARIANT=llvm
west build -p auto -b qemu_x86 -t run tests/lib/c_lib/thrd
```

**Comment by cfriedt** [comment]
Screenshot 2023-09-16 at 3 53 50 PM

**Comment by keith-packard** [comment]
> Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.

Hrm. We should make sure picolibc's header files all come after anything in Zephyr so that Zephyr can replace anything it needs to. Fixing things in picolibc is definitely an option if necessary, of course, but there is a lag between getting that fixed and having it land in an SDK that people are using.

**Comment by cfriedt** [comment]
> > Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.
> 
> Hrm. We should make sure picolibc's header files all come after anything in Zephyr so that Zephyr can replace anything it needs to. Fixing things in picolibc is definitely an option if necessary, of course, but there is a lag between getting that fixed and having it land in an SDK that people are using.

It's a relatively minor change if we ever needed to shuffle contents from `machine/_threads.h` to `threads.h` in zephyr's repo but if the current layout makes it easier to integrate other newlib-based libcs, that is fine with me too 🤷‍♂️

**Comment by nashif** [comment]
> @nashif - does that address your concerns?

Yes, but it would be good to give an overview of this, plans and get more feedback.

**Review by dkalowsk — DISMISSED** [review]
took longer to re-review, but done... thanks for the effort on this!

**Comment by cfriedt** [comment]
> > @nashif - does that address your concerns?
> 
> Yes, but it would be good to give an overview of this, plans and get more feedback.

I've been attending the Arch Review meeting for weeks but this has not been on the agenda. Is there a reason it keeps getting pushed back?

File: tests/lib/c_lib/thrd/testcase.yaml
Comment (aescolar) [inline tests/lib/c_lib/thrd/testcase.yaml:9]: Nowadays this will default to picolibc, so this would need to specifically select CONFIG_MINIMAL_LIBC

File: tests/lib/c_lib/thrd/prj.conf
Comment (aescolar) [inline tests/lib/c_lib/thrd/prj.conf:5]: Nit: maybe it would be a bit nicer if these 2 where in the testcase configuration for minimal libc or an overlay?

**Review by aescolar — DISMISSED** [review]

File: tests/lib/c_lib/thrd/testcase.yaml
Comment (cfriedt) [reply tests/lib/c_lib/thrd/testcase.yaml:9]: Heh.. shows how long this one has been in development 🤣

File: tests/lib/c_lib/thrd/prj.conf
Comment (cfriedt) [reply tests/lib/c_lib/thrd/prj.conf:5]: Maybe even in the testcase.yaml. Good suggestion.

File: lib/libc/common/Kconfig
Comment (galak) [inline lib/libc/common/Kconfig:73]: Worth add a comment here that the POSIX_API `depends` is only because its implemented utilizing that and could be re-implemented using native zephyr api's

**Comment by carlescufi** [comment]
Architecture WG:

- @cfriedt explains that now that we have dynamic thread stacks it is possible to use the C standard to create threads
- @cfriedt explains that this is supposed to work with any of the C standard libraries that we support
- This is pulling in the POSIX API (because it's implemented on top of pthreads), but this can be optimized out later in the cycle. This is not trivial to implement, because the POSIX layer is already doing a lot of what's necessary to implement the API
- @carlescufi: the path of `threads.h` is a bit unusual. Why isn't it in `minimal/include` instead of `common/include`? Why not rely on the standard library-provided headers like we do for the rest? @cfriedt states that the reason is that threads are a new feature and the header is not included in many of those libraries
- @cfriedt states that, in the future, this will support multiple "backends". At that point the infrastructure can be shared between `threads.h` and POSIX so that `threads.h` can be implemented on top of the native Zephyr thread APIs using the common infrastructure.

**Comment by cfriedt** [comment]
@aescolar, @nashif - comments addressed (updated test filters, added Kconfig comment)
@dkalowsk - would you re-ack?

File: tests/lib/c_lib/strerror/prj.conf
Comment (ycsin) [inline tests/lib/c_lib/strerror/prj.conf:2]:
```suggestion
```

`CONFIG_ZTEST_NEW_API` has been removed #63723

File: tests/lib/c_lib/thrd/prj.conf
Comment (ycsin) [inline tests/lib/c_lib/thrd/prj.conf:2]:
```suggestion
```

File: tests/lib/c_lib/strerror/prj.conf
Comment (cfriedt) [reply tests/lib/c_lib/strerror/prj.conf]: Fixed!

File: tests/lib/c_lib/thrd/prj.conf
Comment (cfriedt) [reply tests/lib/c_lib/thrd/prj.conf]: Fixed!

**Comment by cfriedt** [comment]
@aescolar - just added an `arch_exclude` for `posix` to skip the one test that was failing.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/6653568254/job/18079862854?pr=60759

A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.

@dkalowsk, @ycsin, @nashif - should be good to approve once this is green.

**Comment by aescolar** [comment]
> A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.

You can, you just need select picolibc

**Comment by cfriedt** [comment]
> > A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.
> 
> You can, on native_sim. You just need to select picolibc

Ideally, it would work for others too - e.g. minimal, newlib. Maybe not the niche / target-specific libcs.

**Comment by aescolar** [comment]
> Ideally, it would work for others too - e.g. minimal, newlib. Maybe not the niche / target-specific libcs.

With minimal libc it will work also (subject to whatever minimal libc supports). newlib cannot be used with native_sim as that c library is provided by the toolchain.

**Comment by cfriedt** [comment]
* rebased after #63941 to resolve merge conflicts. That was my fault ;-)

**Review by ycsin — DISMISSED** [review]

**Comment by cfriedt** [comment]
@keith-packard - any thoughts?

**Comment by cfriedt** [comment]
@dkalowsk - rebased and all green, if you would like to approve again.

**Review by aescolar — DISMISSED** [review]
The -1 is due to the test exclusion on the posix arch which should not be.
The others are mostly nits.

**Comment by cfriedt** [comment]
* made changes requested by @aescolar

**Review by ycsin — DISMISSED** [review]

File: tests/lib/c_lib/thrd/testcase.yaml
Comment (aescolar) [inline tests/lib/c_lib/thrd/testcase.yaml:12]: The idea was to put this in the common part of the tests definition

Comment (cfriedt) [reply tests/lib/c_lib/thrd/testcase.yaml:12]: Fixed!

**Comment by cfriedt** [comment]
@aescolar, @ycsin, @dkalowsk - it's green. hopefully that works

**Review by ycsin — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by carlescufi — DISMISSED** [review]

**Comment by cfriedt** [comment]
* Rebased again to resolve conflicts
* added module and notls permutations for picolibc to the new testsuites
* added extra configs to provide `rand()` and `srand()` for minimal libc in common testsuite
* reorder test configurations alphabetically
* re-ran entire testsuite locally

Screenshot 2023-11-08 at 12 20 27 PM

Sorry to ask for re-approvals again, @aescolar, @jhedberg , @ycsin , @dkalowsk.

If you have a moment @keith-packard, please take a look.

It's probably safe to merge given there have been no code changes for ~4 months (just rebasing to fix merge conflicts and updating testcase.yaml over and over).

**Review by jhedberg — DISMISSED** [review]

**Review by ycsin — DISMISSED** [review]

**Review by jgl-meta — DISMISSED** [review]
lgtm!

**Review by dkalowsk — DISMISSED** [review]
One minor note I saw on this review, but honestly it can wait for a future code clean up.  Looks good to me!

File: include/zephyr/posix/sys/stat.h
Comment (cfriedt) [reply include/zephyr/posix/sys/stat.h:117]: Yeah, that was a new one - it's because we now support running on `native_sim` (and this is pulled in via picolibc / pthreads headers).

The Great POSIX Header Cleanup (TM) will need to happen soon, I think.

**Comment by cfriedt** [comment]
* rebased (again) to resolve conflicts
* @carlescufi @nashif @stephanosio - can someone please merge this? It really just had merge conflict after merge conflict, has been rebased,  re-approved about a dozen times over the last almost 4 months with practically zero code changes.

**Review by dkalowsk — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Comment by aescolar** [comment]
> * rebased (again) to resolve conflicts
> 
>     * @carlescufi @nashif @stephanosio - can someone please merge this? It really just had merge conflict after merge conflict, has been rebased,  re-approved about a dozen times over the last almost 4 months with practically zero code changes.

@cfriedt note that only admins can override the compliance check failure..

**Review by stephanosio — APPROVED** [review]
LGTM

**Pull request merged**
